### ISSUE TITLE
Problem in setting up multi-node kafka in Swarm mode

### ISSUE THREAD

**Issue by hossein-kshvrz** [wurstmeister/kafka-docker#520]
I am trying to set up a cluster which runs Kafka with multiple servers using docker and in swarm mode. I have created the Swarm properly. Here is my `docker-compose.yml` which is the same as the provided one by contributors [here](https://github.com/wurstmeister/kafka-docker/blob/master/docker-compose-swarm.yml) except in `HOSTNAME_COMMAND` attribute. 
```
version: '3.2'
services:
  zookeeper:
    image: wurstmeister/zookeeper
    ports:
      - "2181:2181"
  kafka:
    image: wurstmeister/kafka:latest
    ports:
      - target: 9094
        published: 9094
        protocol: tcp
        mode: host
    environment:
      HOSTNAME_COMMAND: "ifconfig | grep -A 1 'ens3' | tail -1 | cut -d ':' -f 2 | cut -d ' ' -f 1"
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: INSIDE:PLAINTEXT,OUTSIDE:PLAINTEXT
      KAFKA_ADVERTISED_LISTENERS: INSIDE:/:9092,OUTSIDE://_{HOSTNAME_COMMAND}:9094
      KAFKA_LISTENERS: INSIDE://:9092,OUTSIDE://:9094
      KAFKA_INTER_BROKER_LISTENER_NAME: INSIDE
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
```

I changed that line because when I run `docker info | grep ^Name: | cut -d' ' -f 2` in all of my servers I get an extra line as warning:

    docker info | grep ^Name: | cut -d' ' -f 2
    WARNING: No swap limit support
    master

(I tested with it but it didn't work out so I thought that might be the source of my problem.) 
I use `ifconfig | grep -A 1 'ens3' | tail -1 | cut -d ':' -f 2 | cut -d ' ' -f 1` instead and work with the ip addresses of servers. 
I do `docker stack deploy -c docker-compose.yml wtf` in my master server. Then I do `docker service scale wtf_kafka=3`
Now when I check my service status it is like this:
![Screenshot from 2019-07-14 17-23-28](https://user-images.githubusercontent.com/44027305/61183999-867cb580-a65d-11e9-8646-c71b872b9973.png)
When I check docker processes it is shown that Zookeeper is running on master server.
Now when I use `start-kafka-shell.sh` and go to `/opt/kafka/` I cannot create topic or retrieve topics or perform any other commands! 
![Screenshot from 2019-07-14 17-51-44](https://user-images.githubusercontent.com/44027305/61184238-a661a880-a660-11e9-8756-58c74fe90dd3.jpg)
All I can do is use `--zookeeper` option and create and retrieve topics in that way, which is useless because it is not possible to consume or produce messages in this way.
![Screenshot from 2019-07-14 17-58-49](https://user-images.githubusercontent.com/44027305/61184287-3c95ce80-a661-11e9-839c-acfa999cda94.jpg)

I REALLY NEED SOME HELP! I am trying to deal with this issue for several days. I tested all possible docker compose files and lots of other approaches. Recently I've had nightmares about this. :)))
Please tell me what to do

(Notice: The situation is the same when I don't scale. In that case Zookeeper starts on master and Kafka starts on slave. I use proper ip address in my commands and I test both 9092 and 9094 ports just in case but ...)

**Comment by sscaling** [comment]
With regards to the listener address, this may be useful - https://github.com/wurstmeister/kafka-docker/pull/377/files (However, it still might be an issue as it looks like stderr is polluting the expected output)

I haven't had chance to look at this fully, but might be worth looking through old docker-swarm related issues / questions - https://github.com/wurstmeister/kafka-docker/issues?utf8=%E2%9C%93&q=label%3Adocker-swarm. There are various peoples configs / answers in there.

One thing to check is the broker state stored in zookeeper. If things have restarted you may find that broker IDs don't match hosts etc. If this is the case then a fixed broker ID may help, but then you will need to also configure the logs.dir config and that would largely depend on your persistence strategy.

**Comment by hossein-kshvrz** [comment]
@sscaling
Thanks for your response. I had read all of the issues related to swarm before I posted this issue.
Anyway, I tested your idea about `HOSTNAME_COMMAND` and set it to ``"docker info -f '{{`{{.Swarm.NodeAddr}}`}}'"`` just like the pull request link you provided and it seems that it is working fine! Though I have no clue why other assignments to this variable didn't work. 
You got me out of trouble and I really appreciate it. I was almost desperate since I tried every single instructions available in the Internet and had no more idea how to straighten this problem out. Thanks again!
